src/app/content/highlights/components/CardWrapper.tsx

```diff
@@ -1,30 +1,139 @@
 import Highlighter, { Highlight } from '@openstax/highlighter';
+import { HTMLElement } from '@openstax/types/lib.dom';
 import React from 'react';
-import styled from 'styled-components/macro';
-import theme from '../../../theme';
+import { connect, useSelector } from 'react-redux';
+import styled from 'styled-components';
+import { scrollIntoView } from '../../../domUtils';
+import { AppState } from '../../../types';
+import { assertDefined, assertNotNull, remsToPx } from '../../../utils';
+import * as selectSearch from '../../search/selectors';
+import * as contentSelect from '../../selectors';
+import { cardMarginBottom } from '../constants';
+import { focused } from '../selectors';
 import Card from './Card';
+import { mainWrapperStyles } from './cardStyles';
+import { getHighlightTopOffset, useDebouncedWindowSize } from './cardUtils';
 
-interface Props {
+export interface WrapperProps {
+  hasQuery: boolean;
+  isTocOpen: boolean;
   container: HTMLElement;
   highlighter: Highlighter;
   highlights: Highlight[];
   className?: string;
 }
 
 // tslint:disable-next-line:variable-name
-const Wrapper = ({highlights, className, container, highlighter}: Props) => highlights.length
-  ? <div className={className}>
-      {highlights.map((highlight) => <Card
+const Wrapper = ({highlights, className, container, highlighter}: WrapperProps) => {
+  const element = React.useRef<HTMLElement>(null);
+  const [cardsPositions, setCardsPositions] = React.useState<Map<string, number>>(new Map());
+  const [cardsHeights, setCardsHeights] = React.useState<Map<string, number>>(new Map());
+  const [topOffsets, setTopOffsets] = React.useState<Map<string, number>>(new Map());
+  const [width] = useDebouncedWindowSize();
+  const focusedId = useSelector(focused);
+  const focusedHighlight = React.useMemo(
+    () => highlights.find((highlight) => highlight.id === focusedId),
+    [focusedId, highlights]);
+
+  const onHeightChange = (id: string, ref: React.RefObject<HTMLElement>) => {
+    const height = ref.current && ref.current.offsetHeight;
+    if (cardsHeights.get(id) !== height) {
+      setCardsHeights((previous) => new Map(previous).set(id, height === null ? 0 : height));
+    }
+  };
+
+  const getTopOffsetForHighlight = (highlight: Highlight) => {
+    if (topOffsets.has(highlight.id)) {
+      return assertDefined(topOffsets.get(highlight.id), 'this has to be defined');
+    } else {
+      const offset = assertDefined(
+        getHighlightTopOffset(container, highlight),
+        `Couldn't get top offset for highlight with an id: ${highlight.id}`
+      );
+      setTopOffsets((state) => new Map(state).set(highlight.id, offset));
+      return offset;
+    }
+  };
+
+  const getCardsPositions = React.useCallback(() => {
+    const newPositions: Map<string, number> = new Map();
```

Comment:
It was used before in `const position = cardsPositions.get(focusedHighlight.id);`
instead of `cardsPositions` I was using `getCardsPositions` but it seems I've changed it at some point. I'll check what is better